### PR TITLE
feat(ivy): support i18n & ICU expressions

### PR DESCRIPTION
**Pull request by ocombe**
## PR Type
What kind of change does this PR introduce?

```
[x] Feature
```

## What is the new behavior?
Support for i18n & ICU expressions in ivy.
Implementation of the design https://github.com/angular/angular/pull/26091

## Does this PR introduce a breaking change?
```
[x] No
```

### CODE REVIEW

File: packages/core/test/render3/i18n_spec.ts [hunk 1]
```diff
@@ -7,16 +7,190 @@
...
+    message = `Hello �0�!`;
+    expect(cropMessage(message)).toEqual(message);
+
+    message = `Hello �#2��0��#�!`;
+    expect(cropMessage(message)).toEqual(message);
+
+    // Embedded sub-templates
+    message = `�0� is rendered as: �*3�before�*4�middle�*�after�*�!`;
+    expect(cropMessage(message)).toEqual('�0� is rendered as: �*3��*�!');
+    expect(cropMessage(message, 3)).toEqual('before�*4��*�after');
```
Comment (mhevery) [inline packages/core/test/render3/i18n_spec.ts]: I have changed the encoding rules. See: https://github.com/angular/angular/pull/26091/files#diff-bf70cd3451a2cb2d8e20237ae913bafaR88

File: packages/core/test/render3/i18n_spec.ts [hunk 2]
```diff
@@ -7,16 +7,190 @@
...
+  describe('should generate OpCodes', () => {
+    it('for text', () => {
+      const MSG_DIV = `simple text`;
+      const opCodes = i18nStart(0, MSG_DIV);
+
+      expect(opCodes).toEqual({
+        vars: 1,
+        expandoStartIndex: 0,
+        create: [
+          'simple text', -1 << I18nMutateOpCode.SHIFT_REF | I18nMutateOpCode.AppendChild
```
Comment (mhevery) [inline packages/core/test/render3/i18n_spec.ts]: I think this should be `0` not `-1`. See: https://github.com/angular/angular/pull/26091/files#diff-bf70cd3451a2cb2d8e20237ae913bafaR561 

(same for the remainder of the tests)

File: packages/core/test/render3/i18n_spec.ts [hunk 3]
```diff
@@ -7,16 +7,190 @@
...
+
+    it('for bindings', () => {
+      const MSG_DIV = `Hello �0�!`;
+      const opCodes = i18nStart(0, MSG_DIV);
+
+      expect(opCodes).toEqual({
+        vars: 1,
+        expandoStartIndex: 0,
+        create: [
+          'Hello �0�!', -1 << I18nMutateOpCode.SHIFT_REF | I18nMutateOpCode.AppendChild
```
Comment (mhevery) [inline packages/core/test/render3/i18n_spec.ts]: This should be empty string `''` instead of `Hello �0� `. The reason for this is that empty string is interned and it will marshal faster into DOM. Since we will overwrite it later anyways `document.createText(``)` is faster than `document.createText('Hello �0�')`.

File: packages/core/test/render3/i18n_spec.ts [hunk 4]
```diff
@@ -7,16 +7,190 @@
...
+        ],
+        update: [
+          0b0, // bindings mask
+          4, // if no update, skip 4
+          'Hello ',
+          0, // binding index
+          '!',
+          -1 << I18nUpdateOpCode.SHIFT_REF | I18nUpdateOpCode.Text
```
Comment (mhevery) [inline packages/core/test/render3/i18n_spec.ts]: `-1` is not correct. In this example it should be`0` because `expandoStartIndex === 0`, there fore the text node is stored at that location. 

But this is a bad test because `0` location is where the `i18nStart(0, MSG_DIV)` is stored.

File: packages/core/test/render3/i18n_spec.ts [hunk 5]
```diff
@@ -7,16 +7,190 @@
...
+          'simple text', -1 << I18nMutateOpCode.SHIFT_REF | I18nMutateOpCode.AppendChild
+        ],
+        update: [],
+        icus: null
+      });
+    });
+
+    it('for bindings', () => {
+      const MSG_DIV = `Hello �0�!`;
+      const opCodes = i18nStart(0, MSG_DIV);
```
Comment (mhevery) [inline packages/core/test/render3/i18n_spec.ts]: `i18nStart` should be void. This is because it is public API and we don't want to expose our internals to the outside world. So in tests we should have some other way of retrieving the `TI18n`

File: packages/core/test/render3/i18n_spec.ts [hunk 6]
```diff
@@ -7,16 +7,190 @@
...
+          3, // if no update, skip 3
+          0, // binding index
+          ' is rendered as: ',
+          -1 << I18nUpdateOpCode.SHIFT_REF | I18nUpdateOpCode.Text
+        ],
+        icus: null
+      });
+
+      // OpCodes for the first sub-template
+      opCodes = i18nStart(0, MSG_DIV, 3);
```
Comment (mhevery) [inline packages/core/test/render3/i18n_spec.ts]: `3` is not correct in this location. It should be block id as described here: https://github.com/angular/angular/pull/26091/files#diff-bf70cd3451a2cb2d8e20237ae913bafaR88

File: packages/core/test/render3/i18n_spec.ts [hunk 7]
```diff
@@ -7,16 +7,190 @@
...
+          'middle',
+          -1 << I18nMutateOpCode.SHIFT_REF | I18nMutateOpCode.AppendChild,
+        ],
+        update: [],
+        icus: null
+      });
+    });
+
+    xit('for deleted placeholders', () => {
+      // Original text was: `�0� is rendered as: �*3�before�*4�middle�*�after�*�!`
```
Comment (mhevery) [inline packages/core/test/render3/i18n_spec.ts]: Something is missing in this test. How should the test know what the original text was, or how many placeholders there are.

File: packages/core/src/render3/i18n.ts [hunk 8]
```diff
@@ -6,383 +6,719 @@
...
- * Represents the instructions used to translate the template.
- * Instructions can be a placeholder index, a static text or a simple bit field (`I18nFlag`).
- * When the instruction is the flag `Text`, it is always followed by its text value.
- */
-export type I18nInstruction = number | string;
-/**
- * Represents the instructions used to translate attributes containing expressions.
- * Even indexes contain static strings, while odd indexes contain the index of the expression whose
- * value will be concatenated into the final translation.
+ * Breaks pattern into strings and top level {...} blocks.
```
Comment (mhevery) [inline packages/core/src/render3/i18n.ts]: Just a friendly reminder that this well need useful docs.

File: packages/core/src/render3/i18n.ts [hunk 9]
```diff
@@ -6,383 +6,719 @@
...
+ *
+ * TODO(ocombe): cleanup this code
  */
-export type I18nExpInstruction = number | string;
-/** Mapping of placeholder names to their absolute indexes in their templates. */
-export type PlaceholderMap = {
-  [name: string]: number
-};
-const i18nTagRegex = /{\$([^}]+)}/g;
+function extractParts(pattern: string): (string | /*IcuExpression*/any)[] {
```
Comment (mhevery) [inline packages/core/src/render3/i18n.ts]: need unit tests for this function.

File: packages/core/src/render3/i18n.ts [hunk 10]
```diff
@@ -6,383 +6,719 @@
...
-                index.toString(2).length, 28, `Index ${index} is too big and will overflow`);
+/**
+ * Generate OpCodes that will be used to mutate & update the template for i18n translations
+ *
+ * @param index The index in `TView.data` where the OpCodes will be cached
+ * @param message The translation message
+ * @param subTemplateIndex If we generate the OpCodes for a sub-template then this is the root
+ * template expando index for the sub-template that we want
+ */
+export function i18nStart(index: number, message: string, subTemplateIndex?: number): TI18n {
```
Comment (mhevery) [inline packages/core/src/render3/i18n.ts]: function should have `void` return type.

File: packages/core/src/render3/i18n.ts [hunk 11]
```diff
@@ -6,383 +6,719 @@
...
+            parentIndexStack.push(phIndex);
           }
-          // Add an instruction to remove the expression
-          tmplInstructions.push(index | I18nInstructions.RemoveNode);
-
-          if (index > maxIndex) {
-            maxIndex = index;
+        } else {
+          // It is a closing tag
+          if (value.charAt(0) === '#') {
```
Comment (mhevery) [inline packages/core/src/render3/i18n.ts]: Spec has changed for closing tag.

**Review by mhevery — CHANGES_REQUESTED** [review]

File: packages/core/test/render3/i18n_spec.ts (hunk 7 above)
Comment (ocombe) [reply packages/core/test/render3/i18n_spec.ts]: yes, I wrote this before our discussion with i18nEnd deleting placeholders
that's why the test is `xit`, it wasn't running

File: packages/core/src/render3/interfaces/i18n.ts [hunk 12]
```diff
@@ -190,8 +190,13 @@ export const enum I18nUpdateOpCode {
  *   0b11, 7,
  *   // Concatenate `newValue = 'pre'+lViewData[bindIndex-4]+'in'+lViewData[bindIndex-3]+'post';`.
  *   'pre', -4, 'in', -3, 'post',
+<<<<<<< master
  *   // Update attribute: `elementAttribute(1, 'title', sanitizerFn(newValue));`
  *   1 << SHIFT_REF | Attr, 'title', sanitizerFn,
+=======
+ *   // Update attribute: `elementAttribute(1, 'title', newValue);`
+ *   1 << SHIFT_REF | Attr, 'title',
+>>>>>>> feat(ivy): support i18n & ICU expressions
```
Comment (ocombe) [inline packages/core/src/render3/interfaces/i18n.ts]: woops look like I missed that one, I'll remove the conflict in the next push

File: packages/core/src/render3/i18n.ts [hunk 13]
```diff
@@ -6,398 +6,707 @@
...
+      let value = msgParts[i];
+      if (i & 1) {
+        // Odd indexes are placeholders (elements and sub-templates)
+        if (value.charAt(0) === '/') {
+          // It is a closing tag
+          if (value.charAt(1) === '#') {
+            parentIndex = parentIndexStack.pop() !;
+          }
+        } else {
+          const phIndex = parseInt(value.substr(1), 10);
```
Comment (ocombe) [inline packages/core/src/render3/i18n.ts]: this should probably be `const phIndex = parseInt(value.substr(1), 10) + HEADER_OFFSET`;

File: packages/core/src/render3/i18n.ts [hunk 14]
```diff
@@ -6,398 +6,707 @@
...
+import {addAll} from './util';
+
+const MARKER = `�`;
+const ICU_BLOCK_REGEX = /^\s*(�\d+�)\s*,\s*(select|plural)\s*,/;
+const SUBTEMPLATE_REGEXP = /�\/?\*(\d+:\d+)�/gi;
+const PH_REGEXP = /�(\/?[#*]\d+):?\d*�/gi;
+const BINDING_REGEXP = /�(\d+):?\d*�/gi;
+const ICU_REGEXP = /({\s*�\d+�\s*,\s*\S{6}\s*,[\s\S]*})/gi;
+
+interface IcuExpression {
```
Comment (mhevery) [inline packages/core/src/render3/i18n.ts]: Please add docs explaining: what, how and why.

**Review by mhevery — CHANGES_REQUESTED** [review]
Overall this is looking good! I made some changes here ebe11f299d Please review them to make sure that you agree with them.

One major point is please pay more attention to naming and documentation. At times it is hard to follow as to what is going on. 

Let me know when you have the next set of changes for me to review.

File: packages/core/src/render3/interfaces/icu.ts [hunk 15]
```diff
@@ -0,0 +1,13 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+export const enum IcuType {Plural = 0, Select = 1}
```
[the remark this answers is not in the document]
Comment (ocombe) [reply packages/core/src/render3/interfaces/icu.ts]: actually I just forgot to delete this file :D

File: packages/core/src/render3/i18n.ts [hunk 16]
```diff
@@ -288,492 +473,671 @@ function appendI18nNode(tNode: TNode, parentTNode: TNode, previousTNode: TNode):
...
+let expChangedStackPointer = 0;
+
+/**
+ * Stores the values of the bindings during each update cycle in order to determine if we need to
+ * update the translated nodes.
+ *
+ * @param expression The binding's new value or NO_CHANGE
+ */
+export function i18nExp<T>(expression: T | NO_CHANGE): void {
+  expChangedStack[expChangedStackPointer++] = expression;
```
Comment (mhevery) [inline packages/core/src/render3/i18n.ts]: We should not be having our own stack `expChangedStack` We should just be using `bind` We can retrieve the value using `viewData[bindingIndex - i18nOffset]` also we should be updating the maskBit here. We than don't need `flushExpChangedStack`

File: packages/core/src/render3/i18n.ts [hunk 17]
```diff
@@ -288,492 +473,671 @@ function appendI18nNode(tNode: TNode, parentTNode: TNode, previousTNode: TNode):
...
-      const b1 = idx & 1;
-      // Get the value from the argument vx where x = idx
-      const value = b2 ? (b1 ? v3 : v2) : (b1 ? v1 : v0);
-
-      res += stringify(value);
-    } else {
-      res += instructions[i];
-    }
-  }
+  readUpdateOpCodes(updateOpCodes, icus, viewData);
```
Comment (mhevery) [inline packages/core/src/render3/i18n.ts]: 1. please pass `changeMask` explicitly (rather than rely on global state inside the `readUpdateOpCodes`) function.
2. I think we should have a short circuit `if` here in case `changeMask` is `0` to not bother with anything (most common case)

**Review by mhevery — CHANGES_REQUESTED** [review]
Overall looks good. (I made some minor fixes 6a6c78e0b8) 
Please look at my comments and address them. Ping when ready for last review.

File: packages/core/src/render3/i18n.ts [hunk 18]
```diff
@@ -6,398 +6,707 @@
...
+import {COMMENT_MARKER, ELEMENT_MARKER, I18nMutateOpCode, I18nMutateOpCodes, I18nUpdateOpCode, I18nUpdateOpCodes, IcuType, TIcu} from './interfaces/i18n';
+import {TNode} from './interfaces/node';
+import {HEADER_OFFSET} from './interfaces/view';
+import {addAll} from './util';
+
+const MARKER = `�`;
+const ICU_BLOCK_REGEX = /^\s*(�\d+�)\s*,\s*(select|plural)\s*,/;
+const SUBTEMPLATE_REGEXP = /�\/?\*(\d+:\d+)�/gi;
+const PH_REGEXP = /�(\/?[#*]\d+):?\d*�/gi;
+const BINDING_REGEXP = /�(\d+):?\d*�/gi;
```
[the remark this answers is not in the document]
Comment (ocombe) [reply packages/core/src/render3/i18n.ts]: It's in the design docs. We've discussed this a bit with @AndrewKushnir but we should probably discuss this a bit more.

File: packages/core/src/render3/i18n.ts [hunk 19]
```diff
@@ -6,275 +6,460 @@
...
+    }
+
+    const blocks = extractParts(parts[pos++]) as string[];
+    if (blocks.length) {
+      values.push(blocks);
+    }
+  }
+
+  assertGreaterThan(cases.indexOf('other'), -1, 'Missing key "other" in ICU statement.');
+  // TODO(ocombe): support ICU expressions in attributes, see #21615
```
[the remark this answers is not in the document]
Comment (ocombe) [reply packages/core/src/render3/i18n.ts]: Yes, this is a new feature that was requested by people from Google (they need it for aria labels). It would be nice to add it as the same time but it's not MVP.

File: packages/core/src/render3/i18n.ts [hunk 20]
```diff
@@ -288,492 +473,671 @@ function appendI18nNode(tNode: TNode, parentTNode: TNode, previousTNode: TNode):
...
-          lContainer[RENDER_PARENT] = null;
+    }
+  }
+
+  setIsParent(false);
+
+  return visitedPlaceholders;
+}
+
+function readUpdateOpCodes(
```
[the remark this answers is not in the document]
Comment (ocombe) [reply packages/core/src/render3/i18n.ts]: no because `I18nMutateOpCode.MASK_OPCODE` is different from `I18nUpdateOpCode.MASK_OPCODE`

File: packages/core/src/render3/util.ts [hunk 21]
```diff
@@ -241,3 +241,15 @@ export function getParentInjectorTNode(
...
   return parentTNode;
 }
+
+/**
+ * Equivalent to ES6 spread, add each item to an array.
+ *
+ * @param items The items to add
+ * @param arr The array to which you want to add the items
+ */
+export function addAllToArray(items: any[], arr: any[]) {
```
Comment (AndrewKushnir) [inline packages/core/src/render3/util.ts]: It looks like this function is unused?

File: packages/core/test/render3/i18n_spec.ts [hunk 22]
```diff
@@ -21,1786 +24,1486 @@ const Component: typeof _Component = function(...args: any[]): any {
...
-          //   fin
-          // </ul>
-          template: (rf: RenderFlags, myApp: MyApp) => {
-            if (rf & RenderFlags.Create) {
-              if (!i18n_1) {
-                i18n_1 = i18nMapping(
-                    MSG_DIV_SECTION_1,
-                    [{'START_LI_0': 1, 'START_LI_1': 2, 'START_LI_2': 3}, {'START_LI_1': 0}],
-                    [null, {'EXP_1': 1}], ['START_LI_1']);
-
```
Comment (AndrewKushnir) [inline packages/core/test/render3/i18n_spec.ts]: Do we need this code?

Comment (AndrewKushnir) [inline packages/core/test/render3/i18n_spec.ts]: Do we need these lines?

File: packages/core/src/render3/util.ts (hunk 21 above)
Comment (AndrewKushnir) [inline packages/core/src/render3/util.ts]: We can probably do something like this (that would do the same thing):
```
items.push(...arr);
```

**Review by mhevery — APPROVED** [review]

File: packages/core/src/render3/util.ts (hunk 21 above)
Comment (ocombe) [reply packages/core/src/render3/util.ts]: So apparently using spread is bad because it uses `.apply` which creates more memory pressure, or something like that.

**Review by alxhub — APPROVED** [review]
LGTM from the compiler side - runtime review left to @mhevery

**Comment by mhevery** [comment]
http://test/OCL:221145653:BASE:221145709:1542057665529:fe9a09d2

**Comment by mhevery** [comment]
http://test/OCL:221145653:BASE:221166396:1542064725363:dc3c8482

**Comment by mhevery** [comment]
http://test/OCL:221145653:BASE:221280903:1542130571606:90b96ef2

**Comment by AndrewKushnir** [comment]
Note: 7f9a749 - restoring missing import after rebase
